### PR TITLE
[SCRIPT][API] Add RoBERTa fine-tuning scripts, add BERTClassifier to API

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
- add roberta argument options to the finetune script
- move BERTClassifier to the official API

@hhexiy 

### Essentials ###
- [ ] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by mli** [comment]
Job PR-931/1 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-931/1/index.html

**Comment by mli** [comment]
Job PR-931/2 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-931/2/index.html

**Comment by mli** [comment]
Job PR-931/4 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-931/4/index.html

File: scripts/bert/finetune_classifier.py [hunk 1]
```diff
@@ -512,6 +519,9 @@ def train(metric):
         # inference on dev data
         for segment, dev_data in dev_data_list:
             metric_nm, metric_val = evaluate(dev_data, metric, segment)
+            if metric_val < metric_history[-1][-1] and args.early_stop:
+                should_stop = True
+                logging.info('Early stop at epoch %d', epoch_id)
```
Comment (hhexiy) [inline scripts/bert/finetune_classifier.py]: Might need to have patience, otherwise it can stop very early due to small fluctuation.

Comment (eric-haibin-lin) [reply scripts/bert/finetune_classifier.py]: What's the best practice? Should we have sth like "worse accuracy in two consecutive epochs"?

**Comment by kaonashi-tyc** [comment]
Actually RoBERTa model has a slightly different Classifier structure by default:

https://github.com/pytorch/fairseq/blob/718677ebb044e27aaf1a30640c2f7ab6b8fa8509/fairseq/models/roberta/model.py#L218-L235

Might deserve its own Classifier of sort

**Comment by mli** [comment]
Job PR-931/9 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-931/9/index.html

**Comment by mli** [comment]
Job PR-931/10 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-931/10/index.html

**Comment by mli** [comment]
Job PR-931/11 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-931/11/index.html

**Comment by mli** [comment]
Job PR-931/12 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-931/12/index.html

**Comment by szhengac** [comment]
What is this RoBERT? Any reference?

**Comment by mli** [comment]
Job PR-931/13 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-931/13/index.html

**Comment by eric-haibin-lin** [comment]
https://arxiv.org/abs/1907.11692

**Comment by mli** [comment]
Job PR-931/15 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-931/15/index.html

**Comment by mli** [comment]
Job PR-931/16 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-931/16/index.html

File: src/gluonnlp/data/transforms.py [hunk 2]
```diff
@@ -1221,17 +1221,29 @@ class BERTSentenceTransform:
...
 
-    def __init__(self, tokenizer, max_seq_length, pad=True, pair=True):
+    def __init__(self, tokenizer, max_seq_length, vocab=None, pad=True, pair=True):
         self._tokenizer = tokenizer
         self._max_seq_length = max_seq_length
         self._pad = pad
         self._pair = pair
+        self._vocab = self._tokenizer.vocab if vocab is None else vocab
+        from ..vocab import BERTVocab
+        if isinstance(self._vocab, BERTVocab):
```
Comment (leezu) [inline src/gluonnlp/data/transforms.py]: It may be better to use duck typing here, ie. just check if the vocab defines `cls_token` and `sep_token` attributes. The only advantage of `BERTVocab` over `Vocab` currently is it's `from_sentencepiece` constructor. Once this is factored out we could delete `BERTVocab`. What do you think?

File: src/gluonnlp/data/transforms.py [hunk 3]
```diff
@@ -1221,17 +1221,29 @@ class BERTSentenceTransform:
         Tokenizer for the sentences.
     max_seq_length : int.
         Maximum sequence length of the sentences.
+    vocab : Vocab or BERTVocab
```
Comment (leezu) [inline src/gluonnlp/data/transforms.py]: BERTVocab is a subclass of Vocab. Let's only mention Vocab here. It would be good to document what kind of Vocab special tokens / attributes are expected depending on the type of Bert model (BERT vs RoBERTa).

File: src/gluonnlp/data/transforms.py [hunk 4]
```diff
@@ -1221,17 +1221,29 @@ class BERTSentenceTransform:
...
+        The vocabulary.
```
Comment (leezu) [inline src/gluonnlp/data/transforms.py]: It is not clear that different vocabularies are required/handled for different BERT style models. Let's document that `cls_token`, `sep_token` is used if available and otherwise fallback to `bos_token`, `eos_token`.

To formally specify the expected attributes ( `cls_token`, `sep_token`, etc.) one could (eventually) use Structural subtyping https://mypy.readthedocs.io/en/latest/protocols.html

File: src/gluonnlp/model/bert.py [hunk 5]
```diff
@@ -602,6 +604,174 @@ def __call__(self, inputs, valid_length=None, masked_positions=None):
...
+            Shape (batch_size, num_classes)
+        """
+        # XXX Temporary hack for hybridization as hybridblock does not support None
+        if isinstance(valid_length, list) and len(valid_length) == 0:
+            valid_length = None
+        _, pooler_out = self.bert(inputs, token_types, valid_length)
+        return self.classifier(pooler_out)
+
+class RoBERTaClassifier(HybridBlock):
+    """Model for sentence (pair) classification task with BERT.
```
Comment (kaonashi-tyc) [inline src/gluonnlp/model/bert.py]: The documentation might need be updated that RoBERTa model doesn't have token_types as input

Comment (eric-haibin-lin) [reply src/gluonnlp/model/bert.py]: I didn't include token_types in the documentation for RoBERTa. Where did you see `token_types `?

Comment (kaonashi-tyc) [reply src/gluonnlp/model/bert.py]: @eric-haibin-lin Oh, Oh, nvm. That is a mistake

**Comment by mli** [comment]
Job PR-931/17 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-931/17/index.html
